vcd/resource_vcd_vapp_network.go

```diff
@@ -219,43 +270,169 @@ func resourceVappNetworkRead(d *schema.ResourceData, meta interface{}) error {
 	}
 
 	vAppNetwork := types.VAppNetworkConfiguration{}
+	var networkId string
 	for _, networkConfig := range vAppNetworkConfig.NetworkConfig {
-		if networkConfig.NetworkName == d.Get("name").(string) {
+		networkId, err = govcd.GetUuidFromHref(networkConfig.Link.HREF)
+		if err != nil {
+			return fmt.Errorf("unable to get network ID from HREF: %s", err)
+		}
+		// Check name as well to support old resource ID's which are names and datasources which have names provided by the user
+		if d.Id() == networkId || networkConfig.NetworkName == d.Get("name").(string) {
 			vAppNetwork = networkConfig
+			break
 		}
 	}
 
 	if vAppNetwork == (types.VAppNetworkConfiguration{}) {
-		log.Printf("[DEBUG] Network no longer exists. Removing from tfstate")
-		d.SetId("")
-		return nil
-	}
-
-	d.Set("name", vAppNetwork.NetworkName)
-	d.Set("href", vAppNetwork.HREF)
-	if c := vAppNetwork.Configuration; c != nil {
-		d.Set("fence_mode", c.FenceMode)
-		if c.IPScopes != nil {
-			d.Set("gateway", c.IPScopes.IPScope[0].Gateway)
-			d.Set("netmask", c.IPScopes.IPScope[0].Netmask)
-			d.Set("dns1", c.IPScopes.IPScope[0].DNS1)
-			d.Set("dns2", c.IPScopes.IPScope[0].DNS2)
-			d.Set("dnsSuffix", c.IPScopes.IPScope[0].DNSSuffix)
+		if origin == "resource" {
+			log.Printf("[DEBUG] Network no longer exists. Removing from tfstate")
+			d.SetId("")
+			return nil
+		}
+		return fmt.Errorf("[VAPP network read] didn't find vApp network: %s", d.Get("name").(string))
+	}
+
+	// needs to set for datasource. Do not set always as keep back compatibility when ID was name.
+	if d.Id() == "" {
```

Comment:
Does it mean that for those who had `id=vappNetworkName` we will keep it this way?
I think we have done the switch for `id` to contain real id (not name) in other resources as part of "read" fixes in non major version

Need to decide whether the switch here would make sense.

cc @dataclouder, @lvirbalas

Reply:
this place works only for datasources - we query by name, but Id is real ID.
For back compatibility we don't override id=vappNetworkName to id=id.